File: src/converter/plugins/convertIndexedSignatureDeclaration.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {ifPresent, renderNullable} from "../render.js";
import {InheritanceModifierService, inheritanceModifierServiceKey} from "./InheritanceModifierPlugin.js";

export const convertIndexedSignatureDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isIndexSignatureDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    checkCoverageService?.cover(node)

    const inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)

    const getterInheritanceModifier = inheritanceModifierService?.resolveGetterInheritanceModifier(node, context)
    const setterInheritanceModifier = inheritanceModifierService?.resolveSetterInheritanceModifier(node, context)

    const readonly = node.modifiers?.find(modifier => modifier.kind === ts.SyntaxKind.ReadonlyKeyword)
    readonly && checkCoverageService?.cover(readonly)

    let keyType: string

    if (node.parameters[0].type) {
        keyType = render(node.parameters[0].type)
    } else {
        keyType = "Any? /* type isn't declared */"
    }

    const type = renderNullable(node.type, true, context, render)

    const getter = `
@seskar.js.JsNative
${ifPresent(getterInheritanceModifier, it => `${it} `)}operator fun get(key: ${keyType}): ${type}
    `

    let setter = ""

    if (!readonly) {
        setter = `
@seskar.js.JsNative
${ifPresent(setterInheritanceModifier, it => `${it} `)}operator fun set(key: ${keyType}, value: ${type})
        `
    }


    return `${getter}${ifPresent(setter, it => `\n\n${it}`)}`
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertIndexedSignatureDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.renderNullable
import typescript.SyntaxKind
import typescript.asArray
import typescript.isIndexSignatureDeclaration

val convertIndexedSignatureDeclaration = createPlugin plugin@{ node, context, render ->
    if (!isIndexSignatureDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    checkCoverageService?.cover(node)

    val inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)

    val getterInheritanceModifier = inheritanceModifierService?.resolveGetterInheritanceModifier(node, context)
    val setterInheritanceModifier = inheritanceModifierService?.resolveSetterInheritanceModifier(node, context)

    val readonly = node.modifiers?.asArray()?.find { modifier -> modifier.kind == SyntaxKind.ReadonlyKeyword }
    readonly?.let { checkCoverageService?.cover(it) }

    val firstParameterType = node.parameters.asArray().getOrNull(0)?.type

    val keyType = if (firstParameterType != null) {
        render(firstParameterType)
    } else {
        "Any? /* type isn't declared */"
    }

    val type = renderNullable(node.type, true, context, render)

    val getter = """
@seskar.js.JsNativeGetter
${ifPresent(getterInheritanceModifier) { "$it "}}operator fun get(key: ${keyType}): $type
    """.trim()

    var setter = ""

    if (readonly == null) {
        setter = """
@seskar.js.JsNativeSetter
${ifPresent(setterInheritanceModifier) { "$it "}}operator fun set(key: ${keyType}, value: ${type})
        """.trim()
    }


    "${getter}${ifPresent(setter) { "\n\n${it}" }}"
}
